lib/active_model/serializer/adapter/cached_serializer.rb

```diff
@@ -27,17 +29,21 @@ def fragment_cached?
           @klass._cache_only && !@klass._cache_except || !@klass._cache_only && @klass._cache_except
         end
 
-        def cache_key
+        def cache_key(adapter_instance)
           parts = []
           parts << object_cache_key
+          parts << adapter_instance.class.to_s.demodulize.underscore
           parts << @klass._cache_digest unless @klass._cache_options && @klass._cache_options[:skip_digest]
           parts.join('/')
         end
 
         def object_cache_key
+          return @cached_serializer.object.cache_key if @cached_serializer.object.respond_to? :cache_key
+
+          fail(UndefinedCacheKey, "#{@cached_serializer.object} must define #cache_key, or the cache_key option must be passed into cache on #{@cached_serializer}") if @klass._cache_key.blank?
           object_time_safe = @cached_serializer.object.updated_at
           object_time_safe = object_time_safe.strftime('%Y%m%d%H%M%S%9N') if object_time_safe.respond_to?(:strftime)
-          (@klass._cache_key) ? "#{@klass._cache_key}/#{@cached_serializer.object.id}-#{object_time_safe}" : @cached_serializer.object.cache_key
+          "#{@klass._cache_key}/#{@cached_serializer.object.id}-#{object_time_safe}"
```

Comment:
So that it 

1. prefers the `@cached_serializer.object.cache_key` over `@klass._cache_key`
1. raises an exception if there's no cache key on the cached object nor the serializer

Where previously it would

1. Prefer `@klass._cache_key`
1. When `@klass._cache_key` was not defined, use `@cached_serializer.object.cache_key` even it weren't defined